pkcs8/src/one_asymmetric_key.rs

```diff
@@ -0,0 +1,143 @@
+use der::{BitString, ContextualTo1, Decodable, Encodable, Message};
+
+use crate::{
+    common::{Version, _AttributesStub},
+    AlgorithmIdentifier, Error, Result,
+};
+use core::{convert::TryFrom, fmt};
+
+/// PKCS#8 `OneAsymmetricKey`.
+///
+/// ASN.1 structure containing a [`Version`], an [`AlgorithmIdentifier`], private key
+/// data, and optionally public key data, in an algorithm-specific format.
+///
+/// Described in [RFC 5958 Section 2]:
+///
+/// ```text
+/// OneAsymmetricKey ::= SEQUENCE {
+///     version                   Version,
+///     privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
+///     privateKey                PrivateKey,
+///     attributes            [0] Attributes OPTIONAL,
+///     ...,
+///     [[2: publicKey        [1] PublicKey OPTIONAL ]],
+///     ...
+///   }
+///
+/// Version ::= INTEGER { v1(0), v2(1) } (v1, ..., v2)
+///
+/// PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+///
+/// PrivateKey ::= OCTET STRING
+///
+/// Attributes ::= SET OF Attribute
+///
+/// PublicKey ::= BIT STRING
+/// ```
+///
+/// [RFC 5958 Section 2]: https://datatracker.ietf.org/doc/html/rfc5958#section-2
+#[derive(Clone)]
+pub struct OneAsymmetricKey<'a> {
+    /// PKCS#8 document, only [`Version::V1`] and [`Version::V2`] are valid.
+    pub version: Version,
```

Comment:
I'm a bit torn on whether the version should be explicit/public or hidden.

From my reading of RFC 5958 Section 2, it seems like the version can always be inferred from `public_key`:

```
version identifies the version of OneAsymmetricKey.  If publicKey
     is present, then version is set to v2 else version is set to v1.
```

So perhaps when encoding it should automatically be set to `Version::V2` when `public_key` is present.

When decoding `Version::V2` should always attempt to decode a public key, and treat it as an error if it's missing (apologies, that's not what my code suggestion did, but that does seem like the proper behavior)

Reply:
the RFC says `PublicKey OPTIONAL`, *within* the double braces, so i assume they're optional public keys.

Reply:
Yes, it's optional, but it also says that if it's missing the version should be `Version::V1` (i.e. compatible with RFC5208)

Reply:
Should i remove the version parameter from there (so that impossible situations, i.e. `v1 + pubkey`, cannot happen) and instead infer outgoing version when serializing it, and do the same with incoming version? (`if pubkey then v2 else v1`, (and with deserializing) `if v1 + pubkey: error`)

Reply:
V1 maps to the RFC5208 encoding, which does not have a `public_key` field:

```
     If any items tagged as version
     -- 2 are used, the version must be v2, else the version should be
     -- v1.  When v1, PrivateKeyInfo is the same as it was in [RFC5208].
```

Reply:
Yeah, I'd suggest removing it as it can otherwise be completely automatic, and having an explicit one means the user can get it wrong.